libraries/eosiolib/eosiolib.cpp

```diff
@@ -65,6 +69,75 @@ namespace eosio {
       ds >> params;
    }
 
+   //  considering of runtime efficiency, using a estimated size to save the exact size calculating time
+   const int estimate_param_ids_buff_size = 128;
+   const int estimate_params_buff_size = 256;
+   void set_parameters(const std::vector<std::pair<uint32_t, std::variant<uint16_t, uint32_t, uint64_t>>>& params) {
+      char buff[estimate_params_buff_size];
+      unsigned_int size = params.size();
+      if(params.size() == 0) return;
+      eosio::datastream<char *> ds( buff, sizeof(buff) );
+      ds << size;
+      for(const std::pair<uint32_t, std::variant<uint16_t, uint32_t, uint64_t>> & p : params ){
+         unsigned_int id = p.first;
+         ds << id;
+         std::variant<uint16_t, uint32_t, uint64_t> var = p.second;
+         if(std::holds_alternative<uint16_t>(var)){
+            uint16_t data = std::get<uint16_t>(var);
+            ds << data;
+         } else if (std::holds_alternative<uint32_t>(var) ){
+            uint32_t data = std::get<uint32_t>(var);
+            ds << data;
+         } else {
+            uint64_t data = std::get<uint64_t>(var);
+            ds << data;
+         }
+      }
+      eosio::check( ds.tellp() <= sizeof(buff), "buffer is too small" );
+      set_parameters_packed( buff, ds.tellp() );
+   }
+
+   void get_parameters(const std::vector<uint32_t> & param_ids,  std::vector<std::pair<uint32_t, std::variant<uint16_t, uint32_t, uint64_t>>>& params) {
+      char buff_ids[estimate_param_ids_buff_size];
+      char buff_params[estimate_params_buff_size];
+      unsigned_int id_size = param_ids.size();
+      params.clear();
+      if(param_ids.size() == 0) return;
+      eosio::datastream<char*> id_ds( buff_ids, sizeof(buff_ids) );
+      id_ds << id_size;
+      for(const auto & id : param_ids){
+         unsigned_int _id = id;
+         id_ds << _id;
+      }
+      eosio::check( id_ds.tellp() <= sizeof(buff_ids), "ids buffer is too small" );
+      uint32_t size = get_parameters_packed( buff_ids,  id_ds.tellp(), buff_params, sizeof(buff_params));
+      eosio::check( size <= sizeof(buff_params), "params buffer is too small" );
+      eosio::datastream<const char*> para_ds( buff_params, size );
+      unsigned_int para_size;
+      para_ds >> para_size;
+      eosio::check(id_size == para_size, "pairs returned doesn't match amount of id to query");
+      for(uint32_t i = 0; i < para_size; ++i){
+         unsigned_int _id;
+         std::variant<uint16_t, uint32_t, uint64_t>  var;
+         para_ds >> _id;
+         uint32_t id = _id;
```

Comment:
Could `_id` be just used below, instead of introducing `id`?

Reply:
unsigned_int is a struct, it is not unsigned int, I'm confused with this at first look also. so basically _id and id is different type. _id can pass value to id.  and also _id.value is uint32_t, but if using _id.value,  the code reader can't see the type by first look, so using uint32_t id = _id to give a clear idea to code reader what the type is.

Reply:
I saw `unsigned_int` has an overloaded `==` with `uint32_t` so `_id` might work, but yeah I agree the current code is more explicit.